### PR TITLE
[mlir][gpu] Productize `test-lower-to-nvvm` as `gpu-lower-to-nvvm`

### PR DESCRIPTION
**Pull request by grypp**
The `test-lower-to-nvvm` pipeline serves as the common and proper pipeline for nvvm+host compilation, and it's used across our CUDA integration tests.

This PR updates the `test-lower-to-nvvm` pipeline to `gpu-lower-to-nvvm` and moves it within `InitAllPasses.h`. The aim is to call it from Python, also having a standardize compilation process for nvvm.

### CODE REVIEW

File: mlir/lib/Dialect/GPU/Pipelines/GPUToNVVMPipeline.cpp [hunk 1]
```diff
@@ -164,14 +163,14 @@ void buildLowerToNVVMPassPipeline(OpPassManager &pm,
...
 namespace mlir {
-namespace test {
-void registerTestLowerToNVVM() {
-  PassPipelineRegistration<TestLowerToNVVMOptions>(
-      "test-lower-to-nvvm",
+namespace gpu {
+void registerGPUToNVVMPipeline() {
+  PassPipelineRegistration<GPUToNVVMPipelineOptions>(
+      "gpu-lower-to-nvvm",
       "An example of pipeline to lower the main dialects (arith, linalg, "
```
Comment (ftynse) [inline mlir/lib/Dialect/GPU/Pipelines/GPUToNVVMPipeline.cpp:170]: This is no longer an example, is it?  Could you maybe describe somewhere that is expected to be included into this pipeline in the future (and what is clearly out of scope)?

File: mlir/include/mlir/Dialect/GPU/Pipelines/Passes.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
+//===- Passes.h - GPU NVVM pipeline entry points -----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This header file defines prototypes of all sparse tensor pipelines.
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/GPU/Pipelines/Passes.h:9]: This is not sparse tensor.

File: mlir/include/mlir/Dialect/GPU/Pipelines/Passes.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
+//===- Passes.h - GPU NVVM pipeline entry points -----------*- C++ -*-===//
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/GPU/Pipelines/Passes.h:1]: Nit: 80 cols.

**Review by ftynse — COMMENTED** [review]
The commit title looks scary...

**Review by ftynse — APPROVED** [review]

**Comment by joker-eph** [comment]
I have quite strong concerns with this: I can't see how this is anything else than a "test pipeline": it can only support a very specific set of input IR but the name make it sounds like it is much more generic than it is.

**Comment by ftynse** [comment]
I took it as more aspirational commit to work towards a better flow for the NVVM target. If it's not the case, we should certainly make it clearer in the name.

**Comment by joker-eph** [comment]
Seeing that there are conversion for linalg in this pipeline, it's not clear to me what this aspires to right now.
@grypp can you clarify?

**Comment by grypp** [comment]
I had two key goals: firstly, to use this pipeline via Python, and secondly, having a universal pipeline to compile `nvgpu, memref, vector, scf, vector->nvvm->cubin/ptx`.

Regarding the `createConvertLinalgToLoopsPass`, it appears to generate loops without parallelization. Considering our integration tests, it might be needed. I can try to exclude it and possibly rename the pipeline if it is your concern.

**Comment by grypp** [comment]
> ... it sounds like it is much more generic than it is.

Alternatively, I can try to improve the pipeline and make it generic. But this will require more tests. Don't we want a generic pipeline for nvvm target?

**Comment by joker-eph** [comment]
If this is supposed to be a pipeline that starts from the `gpu` dialect and down, then we should scope it clearly with the input dialects that are supported and the expectation of the input IR. I don't expect to see any linalg there for example.

Right now this pipeline is not documented at all: neither in the header exposing the registration, neither on the website: I'd like to see both :)

**Comment by grypp** [comment]
Sure, I will to make the following changes then
- Exclude linalg
- Add a document header
- Include a document website

**Comment by joker-eph** [comment]
@grypp did you follow up on this?

**Comment by grypp** [comment]
> @grypp did you follow up on this?

Yes, this was on my to-do list. I was out of the office, so submitting the PR took longer than I had anticipated.

**Comment by joker-eph** [comment]
No hurry! As long as it's not lost :)

**Pull request merged**